keps/sig-network/1797-configure-fqdn-as-hostname-for-pods/README.md

```diff
@@ -0,0 +1,762 @@
+  At minimum, you should fill in the "Summary", and "Motivation" sections.
+  These should be easy if you've preflighted the idea of the KEP with the
+  appropriate SIG(s).
+- [ ] **Create a PR for this KEP.**
+  Assign it to people in the SIG that are sponsoring this process.
+- [ ] **Merge early and iterate.**
+  Avoid getting hung up on specific details and instead aim to get the goals of
+  the KEP clarified and merged quickly.  The best way to do this is to just
+  start with the high-level sections and fill out details incrementally in
+  subsequent PRs.
+
+Just because a KEP is merged does not mean it is complete or approved.  Any KEP
+marked as a `provisional` is a working document and subject to change.  You can
+denote sections that are under active debate as follows:
+
+```
+<<[UNRESOLVED optional short context or usernames ]>>
+Stuff that is being argued.
+<<[/UNRESOLVED]>>
+```
+
+When editing KEPS, aim for tightly-scoped, single-topic PRs to keep discussions
+focused.  If you disagree with what is already in a document, open a new PR
+with suggested changes.
+
+One KEP corresponds to one "feature" or "enhancement", for its whole lifecycle.
+You do not need a new KEP to move from beta to GA, for example.  If there are
+new details that belong in the KEP, edit the KEP.  Once a feature has become
+"implemented", major changes should get new KEPs.
+
+The canonical place for the latest set of instructions (and the likely source
+of this file) is [here](/keps/NNNN-kep-template/README.md).
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once
+it is marked `implementable` must be approved by each of the KEP approvers.
+If any of those approvers is no longer appropriate than changes to that list
+should be approved by the remaining approvers and/or the owning SIG (or
+SIG Architecture for cross cutting KEPs).
+-->
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+Giving users the ability to set the hostname field of the kernel to the FQDN of a Pod.
+
+
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+Giving users a way to configure or enforce this feature cluster-wide.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+This proposal gives users the ability to set a pod’s hostname to its FQDN. A new PodSpec field named `fqdnInHostname` 
+will be introduced, with type `*bool`.
+
+The values of `fqdnInHostname` are:
+* `nil` (default): The Linux kernel hostname field ([the nodename field of struct utsname](http://man7.org/linux/man-pages/man2/uname.2.html)) 
```

Comment:
When it comes to impl time we can decide if this defaults to false, which means it might show up and trigger problems with pod hashes (@liggitt) or don't default this, and treat nil and false as the same or don't make it a pointer, and let `omitEmpty` carry the day (does that work for the pod hash problem?)

Reply:
tl;dr: from what I can see, if we don't modify the serialized json output of existing pods we won't trigger kubelet container restarts or controller rollouts.

That means from a restart-safe perspective you can use an `omitempty` field with either of these:
* a `*bool` defaulted to `nil`
* a `bool` defaulted to false

===

There are two places I'm aware of we need to consider:

1. Kubelet

    * The logic the kubelet uses to determine if a container needs restarting is based on the [json serialization of the container manifest](https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/kubelet/container/helpers.go#L98-L108)
    * an `omitempty` field that defaults to the zero value (a `nil *bool` or a `false bool`) wouldn't cause any issues for the kubelet

2. Controllers (ok, I lied... there are lots of controllers)

    * DaemonSet - appears to use [json serialization comparison with the controller revision](https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/controller/daemon/update.go#L271-L305) to determine if a new rollout is required
    * StatefulSet - appears to use json serialization comparison with the controller revision to determine if a new rollout is required
    * Deployment - [compares the hash of the deployment pod template to the child replicaset pod templates](https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/controller/deployment/util/deployment_util.go#L644-L677), so any podspec API change would affect both identically
    * ReplicaSet/ReplicationController - I had a hard time finding if/where these controllers reconcile their pod template with the pod specs of child pods

Reply:
Thanks @thockin and @liggitt , this is a great point and those code links are awesome!